source/op/prod_env_mat_multi_device.cc

```diff
@@ -366,6 +440,44 @@ class ProdEnvMatAOp : public OpKernel {
       if(b_nlist_map) _map_nlist_gpu(nlist, idx_mapping, nloc, nnei);
       deepmd::delete_device_memory(firstneigh);
       #endif //GOOGLE_CUDA
+
+            #if TENSORFLOW_USE_ROCM
+      int * idx_mapping = NULL;
+      int * ilist = NULL, * numneigh = NULL;
+      int ** firstneigh = NULL;
+      deepmd::malloc_device_memory(firstneigh, nloc);
+      int * jlist = NULL;
+      FPTYPE * coord_cpy;
+      int * type_cpy;
+      int frame_nall = nall;
+      int mesh_tensor_size = static_cast<int>(mesh_tensor.NumElements());
+      std::vector<Tensor> tensor_list(7);
+      // prepare coord and nlist
+      _prepare_coord_nlist_gpu_rocm<FPTYPE>(
+          context, &tensor_list[0], &coord, coord_cpy, &type, type_cpy, idx_mapping, 
+          gpu_inlist, ilist, numneigh, firstneigh, jlist, nbor_list_dev,
+          frame_nall, mem_cpy, mem_nnei, max_nbor_size,
+          box, mesh_tensor.flat<int>().data(), mesh_tensor_size, nloc, nei_mode, rcut_r, max_cpy_trial, max_nnei_trial);
+
+      // allocate temp memory, temp memory must not be used after this operation!
+      Tensor int_temp;
+      TensorShape int_shape;
+      int_shape.AddDim(sec_a.size() + nloc * sec_a.size() + nloc);
+      OP_REQUIRES_OK(context, context->allocate_temp(DT_INT32, int_shape, &int_temp));
+      Tensor uint64_temp;
+      TensorShape uint64_shape;
+      uint64_shape.AddDim(nloc * GPU_MAX_NBOR_SIZE * 2);
+      OP_REQUIRES_OK(context, context->allocate_temp(DT_UINT64, uint64_shape, &uint64_temp));
+      array_int = int_temp.flat<int>().data(); 
+      array_longlong = uint64_temp.flat<unsigned long long>().data();
+
+      // launch the gpu(nv) compute function
+      deepmd::prod_env_mat_a_gpu_rocm(
+          em, em_deriv, rij, nlist, 
+          coord, type, gpu_inlist, array_int, array_longlong, max_nbor_size, avg, std, nloc, frame_nall, rcut_r, rcut_r_smth, sec_a);
+      if(b_nlist_map) _map_nlist_gpu_rocm(nlist, idx_mapping, nloc, nnei);
+      deepmd::delete_device_memory(firstneigh);
+      #endif //TENSORFLOW_USE_ROCM
```

Comment:
Can this part merge with the cuda implementation? or introduce the condition only when the function name appears? or considering use template the unify the interface?

Reply:
introduce the condition only when the function name appears is OK,but I didn't adopt this when I transfer the code

Reply:
Apparently you did not. I ask you to do so.